### PR TITLE
Implement dynamic legend (issue #156)

### PR DESCRIPTION
**Pull request by HotLemonJuice**
Related issue #156 

This PR adds a dynamic legend in the normalization mode. 
Boundary = (maxCount - minCount) / (# of color intervals)
[0, 1/6]
[1/6, 2/6]
[2/6, 3/6]
[3/6, 4/6]
[4/6, 5/6]
[5/6+]

Side by side comparison: (left: before, right: after)
screen shot 2017-04-30 at 16 08 08
screen shot 2017-04-30 at 16 09 38

### CODE REVIEW

**Comment by HotLemonJuice** [comment]
@chenlica There are some improvements in the city level, but the improvement on the state level is still minimal. The reason is that sometimes the variance of data is very small. In other words, many states have very similar counts, so they end up with the same color. One way we can fix this entirely is to come up with some algorithm to rank the count so that we are sure to divide them equally. But the trade-off of this approach is that the intervals of the legend might be very strange. For example, it can be something like
[0, 1]
[1, 1.5]
[1.5, 1.8]
[1.8, 10]
[10, 100]
[100, 500]
Any thoughts?

**Comment by chenlica** [comment]
@HotLemonJuice I think if the county-level counts are similar, let it be.  The city-level differences look good!

File: twittermap/public/javascripts/map/controllers.js [hunk 1]
```diff
@@ -447,7 +462,10 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
       }
 
       function setNormalizedCount(geo, r){
-        geo['properties']['count'] = r['count'] / r['population'] * cloudberryConfig.normalizationUpscaleFactor;
+        var normalizedCount = r['count'] / r['population'] * cloudberryConfig.normalizationUpscaleFactor;
+        geo['properties']['count'] = normalizedCount;
+        if(normalizedCount > normalizedCountMax)  // update max to enable dynamic legends
+          normalizedCountMax = normalizedCount;
```
Comment (JavierJia) [inline twittermap/public/javascripts/map/controllers.js]: do we also need to dynamically change it on `count`?

Comment (HotLemonJuice) [inline twittermap/public/javascripts/map/controllers.js]: Are you saying the `'count'` property for each geo? If so, we don't. We only need to keep track of the maximum count, and this is what this line of code is doing.

**Comment by HotLemonJuice** [comment]
So if we are going to use rankings to divide the data, we are sure to have an equal number of polygons in each color interval. But the problem is that it only shows the difference of rankings and it diverges from the message of our original visualization, which was to offer an objective insight into the data.

As Ben Shneiderman (a famous computer scientist in HCI) said:
> The purpose of visualization is insight, not pictures.

I think we shouldn't use ranking, but let the data speak for itself.
@JavierJia @chenlica What do you think?

**Comment by chenlica** [comment]
@HotLemonJuice Seems you have more insights on how to show the data.  Please make a proposal and go ahead to implement it.

**Comment by JavierJia** [comment]
@HotLemonJuice do you have a new proposal or we should proceed to review this change? 😄

**Comment by HotLemonJuice** [comment]
@chenlica I think the current visualization implementation is very good so that's why I said we shouldn't further divide the colors. In other words, let's use the dynamic coloring in this PR. 
 
On the other hand, I have also thought about new visualization implementations. Maybe we can visualize the hashtags because they are topics related to the keyword and they may provide some insight about what people are caring about around this keyword. A good example is a wall of words in which words of higher frequency have larger sizes.

example:
many-eyes

**Comment by HotLemonJuice** [comment]
@JavierJia we should definitely continue reviewing this PR!

**Comment by JavierJia** [comment]
I will try the UDF instruction to add the UDF to AsterixDB docker image so that you can play with the real query. (tonight or tomorrow)

**Comment by chenlica** [comment]
@HotLemonJuice OK, let's use the current way to display colors.  Regarding the new feature, it looks cool.  In general, we want to improve the way to visualize the results without overwhelming the user.  Let's first unify the few features ("count", "normalize", "sentiment") in one sidebar, then think about other improvements.  I may have some other UI tasks for you soon.  I will let you know.

**Comment by HotLemonJuice** [comment]
@JavierJia OK cool!

**Pull request merged**